lua/lazyvim/plugins/extras/coding/mini-ai.lua

```diff
@@ -0,0 +1,139 @@
+local load_textobjects = false
+
+---@param langs table<string>
+---@return table<string>
+local function dedup(langs)
+  -- TODO: move this utility function to a better location
+  ---@type table<string, boolean>
+  local added = {}
+  return vim.tbl_filter(function(lang)
+    if added[lang] then
+      return false
+    end
+    added[lang] = true
+    return true
+  end, langs)
+end
+
+return {
+
+  {
+    "nvim-treesitter/nvim-treesitter",
+    dependencies = {
+      {
+        "nvim-treesitter/nvim-treesitter-textobjects",
+        init = function()
+          -- disable rtp plugin, as we only need its queries for mini.ai
+          -- In case other textobject modules are enabled, we will load them
+          -- once nvim-treesitter is loaded
+          require("lazy.core.loader").disable_rtp_plugin("nvim-treesitter-textobjects")
+          load_textobjects = true
+        end,
+      },
+    },
+    opts = function(_, opts)
+      -- override the default textobjects, set in plugins/treesitter.lua
+      -- when using this extra, the user is expected to add textobjects
+      -- on a need-to-have basis
+      opts.textobjects = nil
+    end,
+    ---@param opts TSConfig
+    config = function(_, opts) -- override config in plugins/treesitter.lua
+      if type(opts.ensure_installed) == "table" then
+        opts.ensure_installed = dedup(opts.ensure_installed)
```

Comment:
@abeldekat I don't believe you need the `dedup` function here as it's already handled internally by [Lazyvim](https://github.com/LazyVim/LazyVim/blob/a72a84972d85e5bbc6b9d60a0983b37efef21b8a/lua/lazyvim/plugins/treesitter.lua#L64-L75)?

Reply:
@dpetka2001,

I appreciate your input!

The link you used points to the old code in `treesitter.lua`, now present in the `dedup` function in both `treesitter.lua` and the `extra`. This explicitly marks the code duplication.
See [here](https://github.com/LazyVim/LazyVim/blob/0c12cc332a7e6f260b8dec361b5474742b432a3a/lua/lazyvim/plugins/treesitter.lua#L3)

In the `extra`, I have to *override* the config function defined in core `treesitter.lua`. Still thinking about a better way...

Best regards!

Reply:
Oh, you're right. Didn't pay attention that you're also overriding the treesitter config in the extras. Essentially you have 2 different treesitter configs depending on if `mini.ai` is enabled or not. I see. Tnx for clarifying.

Reply:
@abeldekat I may be talking rubbish, but here is an idea. What if you left the original code in core `treesitter.lua` like it was, but removing the `init` function from `nvim-treesitter-textobjects`? That ways the code for PERF would never get loaded since `load_textobjects` would always be false. Then in the extras you change in `init` function `load_textobjects` to `true`. Would that help getting rid of having to override the treesitter `config` in the extras as well?

Reply:
@dpetka2001,

Thanks for the suggestion. Inside the `extra`, I would need to access the `load_textobjects` variable defined in `lazyvim.plugins.treesitter`. That is a spec, thus the variable is not within reach, as far as I can see.

However, your comment was very helpful. I found a way to decorate the `config` function in `extra`, thus avoiding any code duplication. Maybe you could inspect the new [commit](https://github.com/LazyVim/LazyVim/pull/1434/commits/cdf8036e7fd2f1cada5c488e5319a26cc71fb33c#diff-597497ffb8a9f32f2e151a7d3ad8b479b0f3999c442d02a55368778282f3f3a7)?

Reply:
@abeldekat 

Thank you for asking my opinion, however I'm going to disappoint you due to lack of programming knowledge. I have read through your code. I believe I understand what it's doing at a pretty good level (for my standards at least). From my understanding it seems good, but you definitely should defer your review request to someone experienced like Folke. I shouldn't be the one reviewing. I can only offer some ideas based on how I can logically deduce some things, but even then I do mistakes like the local variable not being able to be accessed across specs. I really admire the work you put in regardless of the outcome of this PR. And I honestly appreciate it you wanting to include me in something like this, even though you absolutely don't have to. I really hope I had the knowledge to offer you more concrete and valuable feedback, but I'm not at that level. I can only read and understand some code due to my effort to be helpful to this project because I like it. I tried to inspect your PR as best as I could, but I can under no circumstances offer a review opinion on this. If it was something simpler, I could do so, but I know my limits and don't want to offer information on something I can barely understand.
Wish you luck with this :smile:

Reply:
I liked your idea. Thanks, @dpetka2001!